chain/jsonrpc/src/lib.rs

```diff
@@ -571,90 +568,56 @@ impl JsonRpcHandler {
         Ok(RpcProtocolConfigResponse { config_view })
     }
 
-    async fn query(&self, params: Option<Value>) -> Result<Value, RpcError> {
-        let query_request = if let Ok((path, data)) =
-            parse_params::<(String, String)>(params.clone())
-        {
-            // Handle a soft-deprecated version of the query API, which is based on
-            // positional arguments with a "path"-style first argument.
-            //
-            // This whole block can be removed one day, when the new API is 100% adopted.
-            let data = from_base_or_parse_err(data)?;
-            let query_data_size = path.len() + data.len();
-            if query_data_size > QUERY_DATA_MAX_SIZE {
-                return Err(RpcError::server_error(Some(format!(
-                    "Query data size {} is too large",
-                    query_data_size
-                ))));
+    async fn query(
+        &self,
+        request_data: near_jsonrpc_primitives::types::query::RpcQueryRequest,
+    ) -> Result<
+        near_jsonrpc_primitives::types::query::RpcQueryResponse,
+        near_jsonrpc_primitives::types::query::RpcQueryError,
+    > {
+        let query = Query::new(request_data.block_reference, request_data.request);
+        // This match is used here to give backward compatible error message for specific
+        // error variants. Should be refactored once structured errors fully shipped
```

Comment:
:cry: 

I think our only option going forward would be to return both `result` and `error` in the same JSON RPC response. (`result` for backward compatibility, and `error` for future sanity)

Reply:
why couldn't we get rid of `QueryResponseKind::Error` and emulate it here instead?

Reply:
@bowenwang1996 good point.

@khorolets Can we get rid of the `Error` variant and move this handling to a separate backward-compatibility-wrapper called on the `process_request` level? (`serde_json::to_value(query_response)` can moved into this wrapper)